### PR TITLE
cdc: check leadership before subscribing

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

###  What have you changed?

Add `CasualMessage::CaptureChange` for checking leadership before starting to subscribe changes of a region.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.
###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

File: components/cdc/src/endpoint.rs [hunk 1]
```diff
@@ -101,6 +101,7 @@ impl fmt::Debug for Task {
 }
 
 pub struct Endpoint {
+    // TODO: add RaftRouter here.
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs]: Can we add it now?

**Comment by 5kbpers** [comment]
I will submit another PR for reorganize the tests for cdc.

File: components/cdc/src/endpoint.rs [hunk 2]
```diff
@@ -100,10 +101,10 @@ impl fmt::Debug for Task {
     }
 }
 
-pub struct Endpoint {
+pub struct Endpoint<T> {
     capture_regions: HashMap<u64, Delegate>,
     scheduler: Scheduler<Task>,
-    apply_router: ApplyRouter,
+    casual_router: T,
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs]: Could you name it to raft_router? casual_router sounds confusing.

File: components/cdc/src/endpoint.rs [hunk 3]
```diff
@@ -257,6 +269,11 @@ impl Endpoint {
     fn on_region_ready(&mut self, region_id: u64, resolver: Resolver, region: Region) {
         if let Some(delegate) = self.capture_regions.get_mut(&region_id) {
             delegate.on_region_ready(resolver, region);
+            println!("region ready: has_failed({})", delegate.has_failed());
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs]: Please remove this line.

File: components/cdc/src/endpoint.rs [hunk 4]
```diff
@@ -201,30 +202,41 @@ impl Endpoint {
...
+            }
         } else {
-            ApplyTask::Change(ChangeCmd::Snapshot {
+            ChangeCmd::Snapshot {
                 region_id,
                 region_epoch: request.take_region_epoch(),
-                cb: Callback::Read(Box::new(move |resp: ReadResponse<_>| {
+            }
+        };
+        if let Err(e) = self.casual_router.send(
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs]: Could you log the error?

File: components/raftstore/src/store/fsm/peer.rs [hunk 5]
```diff
@@ -481,6 +482,19 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+        if !self.fsm.peer.is_leader() {
+            cb.invoke_with_response(new_error(Error::NotLeader(
+                self.region_id(),
+                self.fsm.peer.get_peer_from_cache(self.fsm.peer.leader_id()),
+            )));
+            return;
+        }
+        self.ctx
+            .apply_router
+            .schedule_task(self.region_id(), ApplyTask::Change { cmd, cb })
```
Comment (MyonKeminta) [inline components/raftstore/src/store/fsm/peer.rs]: Is it possible that the leader changes after this but before the cmd is executed?

Comment (5kbpers) [reply components/raftstore/src/store/fsm/peer.rs]: Yes, but role change events will be caught by the cdc observer.

**Pull request merged**